### ISSUE TITLE
XrdCl calls close-file callback twice

### ISSUE THREAD

**Issue by bbockelm** [xrootd/xrootd#488]
See:

https://github.com/cms-sw/cmssw/issues/18102

CMS is seeing the xrootd client (`XrdCl`) invoke the callback twice.  Not 100% sure, but seems to occur when the client has run read recovery and is still hitting timeouts.

This is with xrootd-4.5.0.

**Comment by simonmichal** [comment]
Hi Brian,

Thanks for reporting it.

Could you see if it can be reproduced with xrootd-4.6.0?
A similar problem has been fixed in #483 (which is included in 4.6.0)

In particular, we have seen that if the read time-out kicks in and the stream is broken the request and its handler due to an error are owned both by the outgoing and incoming queue. This triggers all sorts of errors. 
I'm suspecting that in your case a second read time-out kicked in and the handler and its request were transferred for a second time to the incoming queue (from what I see in the back trace there were two read time-outs, right?). Then, the thread-pool simply invoked both handlers/callbacks.

Cheers,
Michal

**Comment by bbockelm** [comment]
We likely need a 4.6.1 tag (or pre-release tag) first before testing.  Can you do that?

**Comment by simonmichal** [comment]
We will have 4.6.1 release candidate within few days.

Can you not use the 4.6.0-8 that is now in EPEL stable?

Michal

**Comment by bbockelm** [comment]
CMS aims to build on git tags, I believe.  If you tag the release candidate, it's easiest.

**Comment by bbockelm** [comment]
@simonmichal - we applied the patch from #483 and Xrootd client crashes become significantly more frequent.

It actually appears to be related to this invalid memory handling:

https://github.com/cms-sw/cmssw/issues/14952#issuecomment-243616732

and seems to usually be prefixed by the following:

```
[2016-08-30 16:52:52.468355 -0400][Error  ][XRootDTransport   ] Message 0xf0270d30, stream [1, 0] is a response that we're no longer interested in (timed out)
```

I.e., it seems that the same message is getting processed twice - the first time it is processed, it is deleted.  The second time, it segfaults the application.

This is becoming a blocker for our physics release for the year, so help would be appreciated.

**Comment by simonmichal** [comment]
Hi Brian,

This seems to be a different problem than the original one.

Previously, it was the FileStateHandler object that got deleted twice:

1) There was a read timeout for a request that was in the inQueue 
   (waiting for a response), this triggered the CloseHandler callback,
   which in turn triggered your routine DelayedClose, which deletes the
   FileStateHandler.

2) There was a second read timeout for the same request that was handled in 1).
   So the same things happend and we got a double delete. The bottom line is
   the request must have been reinserted into the inQueue.

Since we have seen something similar I pointed you to #483


Now, in this case, we see a double delete of the incoming response:

1) A new response comes in and a new Message object gets allocated:

https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClAsyncSocketHandler.cc#L585

   The AsyncSocketHandler holds a pointer to this message: pIncoming.
   Then the Message is being passed to the Stream::OnIncoming(), and in here 
   it is being passed further to the XRootDTransport::MessageReceived().

   In the MessageReceived we check if the SID of the message is in the Timed-Out SIDs set,
   and if yes we remove it from the set and delete the message:

https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClXRootDTransport.cc#L1128

   According to the memcheck output that's what happend.

   The MessageReceived routine returns DigestMsg, so the Stream::OnIncoming knows that the
   message was deleted and it returns:

https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClStream.cc#L465


2) Then from the memcheck output we now that the same message object has been deleted for 
   a second time in XRootDMsgHandler::Process method, the callstack looks like that:

   JobManager::RunJobs() -> HandleIncMsgJob::Run() -> XRootDMsgHandler::Process

   Now the really peculiar thing is that we are creating the HandleIncMsgJob only in one place:

https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClStream.cc#L509

   The HandleIncMsgJob is being created (it holds pointer to the message) and submitted 
   to the thread-pool.

   Note that this happens also in the Stream::OnIncoming, and it happens after the return 
   statement we reached in 1), so this means the message was not submitted to the 
   thread-pool in 1). It seems that the Stream::OnIncoming has been invoked for a second
   time with the same message. But the Stream::OnIncoming is only being called from the
   AsyncSocketHandler, and the AsyncSocketHandler resets the pointer to pIncoming after
   calling Stream::OnIncoming:

https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClAsyncSocketHandler.cc#L665

   So right now I cannot see why this happened and I'm really puzzled, 
   I need more time to investigate. 

Cheers,
Mcihal

**Comment by bbockelm** [comment]
Hi Michal,

In parallel, we are blindly upgrading to the `4.6.1-rc1` that you just tagged.  We also saw this in `4.5.0`.

Can you open a separate ticket?

Brian

**Comment by abh3** [comment]
Hi Brian,

This will be done. No RPM wil be generated for RC1 and a patch will be 
developed for RC2. The reason that we even got this far is because we 
thought the problem was fixed until, of course, the posting about the 
persistence of the problem right after RC1 was tagged. Even if we released 
RC1 it was still only a candidate (albiet a faled one).

Andy

**Comment by simonmichal** [comment]
Hi Brian,

Sure I will create a separate ticket! I just noticed your message after tagging RC1. 
Unfortunately the RC1 just died a sudden death.

Cheers,
Michal

**Comment by bbockelm** [comment]
No problem!  Having an RC helps us progress on other items downstream - we can still test 90% of Xrootd even if there's a known issue!

**Comment by simonmichal** [comment]
Is there a chance you guys could build with -O0 and then try to reproduce the problem (running with memcheck)?

In the backtrace some line numbers are massively shifted.

Cheers,
Michal

**Comment by xrootd-dev** [comment]
Hmmm, if the lines are massively shifted it’s because the code you have is not the same code they have. Optimization nowadays is pretty goof at keeping the line numbers relevant. As I recall, cms does have additional patches they apply to the xrootd which may be why the lines are shifted.

Andy

**Comment by simonmichal** [comment]
Hmm not necessarily, if you have a look at the crash reported by Andrea in #463 the backtrace didn't make sense at all, once I compiled with -DCMAKE_BUILD_TYPE=Debug it all started to be clear ;-)

**Comment by simonmichal** [comment]
I created a separate issue to follow up on the double delete of an incoming response: #499

**Comment by simonmichal** [comment]
I suppose I can close this one?

**Comment by bbockelm** [comment]
Yes, I believe this is covered.  Thanks!
